File: src/main/java/io/github/gravetii/controller/GamePlayStyler.java
package io.github.gravetii.controller;

import javafx.animation.*;
import javafx.scene.image.ImageView;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.Pane;
import javafx.util.Duration;

import java.util.Collection;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

public class GamePlayStyler {

  private final GridPane gamePane;
  private final Collection<ImageView> imgViews;
  private final LinkedList<ImageView> seq;
  private final PauseTransition pauser;
  private final SequentialTransition sequencer;

  private Timeline timeline;
  private int gridRotationCount = 0;

  public GamePlayStyler(GridPane gamePane, Collection<ImageView> imgViews) {
    this.gamePane = gamePane;
    this.imgViews = imgViews;
    this.seq = new LinkedList<>();
    this.timeline = new Timeline();
    this.pauser = new PauseTransition(Duration.millis(200));
    this.pauser.setOnFinished((e) -> this.invalidate());
    this.sequencer = new SequentialTransition();
  }

  public void invalidate() {
    this.revert();
    this.reset();
  }

  public void forCorrectWord() {
    this.revert();
    this.applyRotateTransition();
    this.reset();
  }

  public void forIncorrectWord() {
    this.revert();
    this.applyFadeTransition();
    this.reset();
  }

  public void forRepeatedWord() {
    this.revert();
    this.applyScaleTransition();
    this.reset();
  }

  public void forLastInvalidClick(ImageView imgView) {
    this.revert(imgView);
    this.truncate();
  }

  public void forValidClick(ImageView imgView) {
    imgView.getStyleClass().add("custom-img-view-click");
    this.seq.add(imgView);
  }

  private void revert(ImageView imgView) {
    imgView.getStyleClass().remove("custom-img-view-click");
  }

  private void revert() {
    this.seq.forEach(this::revert);
  }

  public void rotate() {
    this.rotateGamePane();
    this.rotateGrid();
  }

  public void rotateOnEnd() {
    while (this.gridRotationCount != 0) this.rotateGrid();
  }

  private void rotate(ImageView imgView) {
    RotateTransition transition = new RotateTransition(Duration.millis(65), imgView);
    transition.setByAngle(360);
    transition.setCycleCount(4);
    transition.play();
  }

  private void applyRotateTransition() {
    this.seq.forEach(this::rotate);
  }

  private void fade(ImageView imgView) {
    FadeTransition transition = new FadeTransition(Duration.millis(50), imgView);
    transition.setFromValue(1.0);
    transition.setToValue(0.3);
    transition.setCycleCount(6);
    transition.setAutoReverse(true);
    transition.play();
  }

  private void applyFadeTransition() {
    this.seq.forEach(this::fade);
  }

  private void scale(ImageView imgView) {
    ScaleTransition transition = new ScaleTransition(Duration.millis(50), imgView);
    transition.setByX(0.2);
    transition.setByY(0.2);
    transition.setCycleCount(6);
    transition.setAutoReverse(true);
    transition.play();
  }

  private void applyScaleTransition() {
    this.seq.forEach(this::scale);
  }

  private void endTransition(ImageView imgView) {
    ScaleTransition scale = new ScaleTransition(Duration.millis(200));
    scale.setByX(0.5);
    scale.setByY(0.5);
    scale.setCycleCount(2);
    scale.setAutoReverse(true);
    RotateTransition rotate = new RotateTransition(Duration.millis(200));
    rotate.setByAngle(360);
    rotate.setCycleCount(2);
    ParallelTransition transition = new ParallelTransition(imgView, rotate, scale);
    transition.play();
  }

  public void applyEndTransition() {
    this.imgViews.forEach(this::endTransition);
  }

  private void truncate() {
    this.seq.pollLast();
  }

  private void reset() {
    this.seq.clear();
  }

  public void revisit(List<ImageView> imgViews) {
    this.timeline.stop();
    this.invalidate();
    Iterator<ImageView> itr = imgViews.iterator();
    KeyFrame keyframe = new KeyFrame(Duration.millis(300), (e) -> this.forValidClick(itr.next()));
    this.timeline = new Timeline(keyframe);
    this.timeline.setOnFinished((e) -> this.pauser.play());
    this.timeline.setCycleCount(imgViews.size());
    this.timeline.play();
  }

  private Pane[][] fetchPanes() {
    int c = 0;
    Pane[][] panes = new Pane[4][4];
    for (int i = 0; i < 4; ++i) {
      for (int j = 0; j < 4; ++j) {
        panes[i][j] = (Pane) this.gamePane.getChildren().get(c++);
      }
    }

    return panes;
  }

  private void rearrange() {
    Pane[][] panes = this.fetchPanes();
    for (int x = 0; x < 2; x++) {
      for (int y = x; y < 3 - x; y++) {
        Pane temp = panes[y][3 - x];
        panes[y][3 - x] = panes[x][y];
        panes[x][y] = panes[3 - y][x];
        panes[3 - y][x] = panes[3 - x][3 - y];
        panes[3 - x][3 - y] = temp;
      }
    }

    for (int i = 0; i < 4; ++i) {
      for (int j = 0; j < 4; ++j) {
        Pane pane = panes[i][j];
        GridPane.setRowIndex(pane, i);
        GridPane.setColumnIndex(pane, j);
      }
    }
  }

  private void rotateGrid() {
    this.rearrange();
    if (++this.gridRotationCount % 4 == 0) this.gridRotationCount = 0;
  }

  private void rotateGamePane() {
    RotateTransition gridTransition = new RotateTransition(Duration.millis(200), this.gamePane);
    gridTransition.setByAngle(360);
    gridTransition.setCycleCount(1);
    this.sequencer.getChildren().add(gridTransition);
    imgViews.forEach(
        imgView -> {
          RotateTransition imgViewTransition = new RotateTransition(Duration.millis(20), imgView);
          imgViewTransition.setByAngle(90);
          imgViewTransition.setCycleCount(2);
          imgViewTransition.setAutoReverse(true);
          this.sequencer.getChildren().add(imgViewTransition);
        });

    this.sequencer.play();
    this.sequencer.setOnFinished((e) -> this.sequencer.getChildren().clear());
  }

  public void applyOnEnd() {
    while (this.gridRotationCount != 0) this.rotateGrid();
  }
}

File: src/main/kotlin/io/github/gravetii/controller/GamePlayStyler.kt
package io.github.gravetii.controller

import javafx.animation.*
import javafx.scene.image.ImageView
import javafx.scene.layout.GridPane
import javafx.scene.layout.Pane
import javafx.util.Duration

class GamePlayStyler(private val gamePane: GridPane, private val imgViews: MutableCollection<ImageView>) {

    private val seq: ArrayList<ImageView> = arrayListOf()
    private var timeline = Timeline()
    private var gridRotationCount = 0
    private val pauser = PauseTransition(Duration.millis(200.0))
    private val panes = fetchPanes()
    private var sequencer = SequentialTransition()

    init {
        pauser.setOnFinished { invalidate() }
    }

    private fun ImageView.revert() = styleClass.remove("custom-img-view-click")

    private fun revert() = seq.forEach { it.revert() }

    private fun reset() = seq.clear()

    private fun truncate() = seq.removeLast()

    private fun applyRotateTransition() {
        val transition = ParallelTransition()
        seq.forEach {
            val rotate = RotateTransition(Duration.millis(65.0), it)
            rotate.byAngle = 360.0
            rotate.cycleCount = 4
            transition.children.add(rotate)
        }

        transition.play()
    }

    private fun applyFadeTransition() {
        val transition = ParallelTransition()
        seq.forEach {
            val fade = FadeTransition(Duration.millis(50.0), it)
            fade.fromValue = 1.0
            fade.toValue = 0.3
            fade.cycleCount = 6
            fade.isAutoReverse = true
            transition.children.add(fade)
        }

        transition.play()
    }

    private fun applyScaleTransition() {
        val transition = ParallelTransition()
        seq.forEach {
            val scale = ScaleTransition(Duration.millis(50.0), it)
            scale.byX = 0.2
            scale.byY = 0.2
            scale.cycleCount = 6
            scale.isAutoReverse = true
            transition.children.add(scale)
        }

        transition.play()
    }

    fun applyEndTransition() {
        val transition = ParallelTransition()
        imgViews.map {
            val scale = ScaleTransition(Duration.millis(100.0))
            scale.byX = 0.5
            scale.byY = 0.5
            scale.cycleCount = 4
            scale.isAutoReverse = true
            val translate = TranslateTransition(Duration.millis(100.0))
            translate.byX = 0.5
            translate.byY = 0.5
            translate.cycleCount = 4
            translate.isAutoReverse = true
            transition.children.add(ParallelTransition(it, scale, translate))
        }

        transition.play()
    }

    private fun fetchPanes(): Array<Array<Pane>> {
        val itr = gamePane.children.iterator()
        return (0 until 4).map { (0 until 4).map { itr.next() as Pane }.toTypedArray() }
            .toTypedArray()
    }

    private fun rearrange() {
        for (i in 0 until 2) {
            for (j in i until 3 - i) {
                val pane = panes[i][j]
                panes[i][j] = panes[3 - j][i]
                panes[3 - j][i] = panes[3 - i][3 - j]
                panes[3 - i][3 - j] = panes[j][3 - i]
                panes[j][3 - i] = pane
            }
        }

        for (i in 0 until 4) {
            for (j in 0 until 4) {
                val pane = panes[i][j]
                GridPane.setRowIndex(pane, i)
                GridPane.setColumnIndex(pane, j)
            }
        }
    }

    private fun rotateGrid() {
        rearrange()
        if (++gridRotationCount % 4 == 0) gridRotationCount = 0
    }

    private fun rotateGamePane() {
        val gridTransition = RotateTransition(Duration.millis(200.0), gamePane)
        gridTransition.byAngle = 360.0
        gridTransition.cycleCount = 1
        sequencer.children.add(gridTransition)
        imgViews.shuffled().forEach {
            val imgViewTransition = RotateTransition(Duration.millis(20.0), it)
            imgViewTransition.byAngle = 90.0
            imgViewTransition.cycleCount = 2
            imgViewTransition.isAutoReverse = true
            sequencer.children.add(imgViewTransition)
        }

        sequencer.setOnFinished { sequencer.children.clear() }
        sequencer.play()
    }

    fun revisit(imgViews: List<ImageView>) {
        timeline.stop()
        invalidate()
        val itr = imgViews.iterator()
        val keyframe = KeyFrame(Duration.millis(300.0), { forValidClick(itr.next()) })
        timeline = Timeline(keyframe)
        timeline.cycleCount = imgViews.size
        timeline.setOnFinished { pauser.play() }
        timeline.play()
    }

    fun invalidate() {
        revert()
        reset()
    }

    fun forValidClick(imgView: ImageView) {
        imgView.styleClass.add("custom-img-view-click")
        seq.add(imgView)
    }

    fun forLastInvalidClick(imgView: ImageView) {
        imgView.revert()
        truncate()
    }

    fun forRepeatedWord() {
        revert()
        applyScaleTransition()
        reset()
    }

    fun forIncorrectWord() {
        revert()
        applyFadeTransition()
        reset()
    }

    fun forCorrectWord() {
        revert()
        applyRotateTransition()
        reset()
    }

    fun rotate() {
        rotateGamePane()
        rotateGrid()
    }

    fun rotateOnEnd() {
        invalidate()
        while (gridRotationCount != 0) rotateGrid()
    }

}